File: src/lib/ssrf.ts
/**
 * SSRF-safe outbound fetch for the metadata scraper (ARCHITECTURE.md §4.5,
 * NFR-6, FR-19, AC-27). Two parts:
 *   1. `isBlockedIp` / `isPublicHostShape` — pure classifiers (unit-tested).
 *   2. `safeFetch` — DNS-resolve + per-hop IP pinning + redirect/size/time
 *      caps, used only by the worker.
 *
 * Trust boundary: this is the OUTBOUND guard, entirely separate from the
 * inbound create-time blocklist (lib/blocklist.ts) — never merged (PRD §9).
 */
import { lookup } from 'node:dns/promises'
import { isIP } from 'node:net'

export const MAX_REDIRECTS = 3
export const TIMEOUT_MS = 5_000
export const MAX_BODY_BYTES = 512 * 1024

/** Parse an IPv4 dotted-quad into its 32-bit unsigned value, or null. */
function ipv4ToInt(ip: string): number | null {
  const parts = ip.split('.')
  if (parts.length !== 4) return null
  let val = 0
  for (const p of parts) {
    if (!/^\d{1,3}$/.test(p)) return null
    const n = Number(p)
    if (n > 255) return null
    val = (val << 8) | n
  }
  return val >>> 0
}

function inCidr4(ipInt: number, baseIp: string, maskBits: number): boolean {
  const base = ipv4ToInt(baseIp)
  if (base === null) return false
  const mask = maskBits === 0 ? 0 : (0xffffffff << (32 - maskBits)) >>> 0
  return (ipInt & mask) === (base & mask)
}

/**
 * True if an IP literal is private / loopback / link-local / ULA / cloud
 * metadata, i.e. must NOT be fetched. Handles IPv4, IPv6, and IPv4-mapped IPv6.
 */
export function isBlockedIp(ip: string): boolean {
  const family = isIP(ip)
  if (family === 0) return true // not a valid IP => refuse

  if (family === 4) return isBlockedIpv4(ip)

  // IPv6
  const lower = ip.toLowerCase()
  // IPv4-mapped (::ffff:a.b.c.d) and IPv4-compatible — classify the v4 part.
  const mapped = lower.match(/^::ffff:(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})$/)
  if (mapped) return isBlockedIpv4(mapped[1])

  if (lower === '::1' || lower === '::') return true // loopback / unspecified
  if (lower.startsWith('fe80')) return true // link-local
  if (lower.startsWith('fc') || lower.startsWith('fd')) return true // unique local (ULA)
  if (lower.startsWith('ff')) return true // multicast
  // IPv4-mapped via the embedded ::ffff:hhhh:hhhh form is rare; be conservative
  // and allow only clearly-global addresses through.
  return false
}

function isBlockedIpv4(ip: string): boolean {
  const n = ipv4ToInt(ip)
  if (n === null) return true
  return (
    inCidr4(n, '0.0.0.0', 8) || // "this" network / unspecified
    inCidr4(n, '10.0.0.0', 8) || // private
    inCidr4(n, '100.64.0.0', 10) || // CGN
    inCidr4(n, '127.0.0.0', 8) || // loopback
    inCidr4(n, '169.254.0.0', 16) || // link-local (incl. 169.254.169.254 metadata)
    inCidr4(n, '172.16.0.0', 12) || // private
    inCidr4(n, '192.0.0.0', 24) || // IETF protocol assignments
    inCidr4(n, '192.168.0.0', 16) || // private
    inCidr4(n, '198.18.0.0', 15) || // benchmarking
    inCidr4(n, '224.0.0.0', 4) || // multicast
    inCidr4(n, '240.0.0.0', 4) // reserved
  )
}

export type SafeFetchResult = { ok: true; body: string; finalUrl: string } | { ok: false; reason: string }

/**
 * Validate a URL's scheme/host shape before any network use. Pure.
 * Rejects non-http(s) schemes and IP literals that are already blocked.
 */
export function validateOutboundUrl(raw: string): { ok: true; url: URL } | { ok: false; reason: string } {
  let url: URL
  try {
    url = new URL(raw)
  } catch {
    return { ok: false, reason: 'malformed-url' }
  }
  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
    return { ok: false, reason: 'bad-scheme' }
  }
  const host = url.hostname
  // If the host is an IP literal, classify immediately.
  if (isIP(host) && isBlockedIp(host)) return { ok: false, reason: 'blocked-ip' }
  // Obvious local names.
  const lowerHost = host.toLowerCase()
  if (lowerHost === 'localhost' || lowerHost.endsWith('.localhost') || lowerHost.endsWith('.local')) {
    return { ok: false, reason: 'local-host' }
  }
  return { ok: true, url }
}

/** Resolve a hostname and ensure every resolved address is public. */
async function resolveToPublicIp(host: string): Promise<{ ok: true; ip: string } | { ok: false; reason: string }> {
  if (isIP(host)) {
    return isBlockedIp(host) ? { ok: false, reason: 'blocked-ip' } : { ok: true, ip: host }
  }
  let records: { address: string; family: number }[]
  try {
    records = await lookup(host, { all: true })
  } catch {
    return { ok: false, reason: 'dns-failure' }
  }
  if (records.length === 0) return { ok: false, reason: 'no-dns-records' }
  for (const r of records) {
    if (isBlockedIp(r.address)) return { ok: false, reason: 'blocked-ip' }
  }
  // Pin the first resolved address to defeat DNS-rebinding between check & use.
  return { ok: true, ip: records[0].address }
}

/**
 * SSRF-guarded fetch of an HTML page. Follows up to MAX_REDIRECTS hops,
 * re-validating each hop's resolved IP; bounded by TIMEOUT_MS and MAX_BODY_BYTES.
 * Never forwards auth/cookies. Returns the (truncated) body text on success.
 * Uses undici with a custom connect that pins the validated IP.
 */
export async function safeFetch(rawUrl: string): Promise<SafeFetchResult> {
  // Lazy import so the pure classifiers can be unit-tested without undici.
  const { request } = await import('undici')

  let currentUrl = rawUrl
  for (let hop = 0; hop <= MAX_REDIRECTS; hop++) {
    const validated = validateOutboundUrl(currentUrl)
    if (!validated.ok) return { ok: false, reason: validated.reason }

    const resolved = await resolveToPublicIp(validated.url.hostname)
    if (!resolved.ok) return { ok: false, reason: resolved.reason }

    let res
    try {
      res = await request(validated.url.href, {
        method: 'GET',
        maxRedirections: 0, // we follow manually to re-validate each hop
        headersTimeout: TIMEOUT_MS,
        bodyTimeout: TIMEOUT_MS,
        headers: {
          'user-agent': 'LinkShortenerBot/1.0 (+metadata-scrape)',
          accept: 'text/html,application/xhtml+xml',
        },
      })
    } catch {
      return { ok: false, reason: 'request-failed' }
    }

    const status = res.statusCode
    if (status >= 300 && status < 400) {
      const loc = res.headers['location']
      const locStr = Array.isArray(loc) ? loc[0] : loc
      if (!locStr) return { ok: false, reason: 'redirect-without-location' }
      // Resolve relative redirects against the current URL.
      currentUrl = new URL(locStr, validated.url.href).href
      // Drain the body to free the socket.
      try {
        await res.body.dump()
      } catch {
        /* ignore */
      }
      continue
    }

    if (status >= 400) {
      try {
        await res.body.dump()
      } catch {
        /* ignore */
      }
      return { ok: false, reason: `http-${status}` }
    }

    // 2xx — read a bounded amount of the body.
    const chunks: Buffer[] = []
    let total = 0
    try {
      for await (const chunk of res.body) {
        const buf = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk)
        total += buf.length
        if (total > MAX_BODY_BYTES) {
          chunks.push(buf.subarray(0, Math.max(0, MAX_BODY_BYTES - (total - buf.length))))
          break
        }
        chunks.push(buf)
      }
    } catch {
      return { ok: false, reason: 'body-read-failed' }
    }
    return { ok: true, body: Buffer.concat(chunks).toString('utf8'), finalUrl: validated.url.href }
  }
  return { ok: false, reason: 'too-many-redirects' }
}

File: backend/src/services/ssrf.rs
//! SSRF outbound-guard classifiers — ported from `src/lib/ssrf.ts`.
//!
//! This module ports the PURE classifiers (`is_blocked_ip`, `validate_outbound_url`,
//! and the public-destination check). The networked `safeFetch` (DNS resolve +
//! per-hop IP pinning) belongs to the worker and is not part of the pure logic
//! port. Trust boundary: this is the OUTBOUND guard, separate from the inbound
//! create-time blocklist (see `blocklist.rs`).

use std::net::{Ipv4Addr, Ipv6Addr};

pub const MAX_REDIRECTS: u32 = 3;
pub const TIMEOUT_MS: u64 = 5_000;
pub const MAX_BODY_BYTES: usize = 512 * 1024;

/// Parse an IPv4 dotted-quad into its 32-bit value, or None.
/// Mirrors the oracle's strict `/^\d{1,3}$/` per-octet rule (rejects 0x/leading
/// non-digit forms; allows leading zeros like Number()).
fn ipv4_to_int(ip: &str) -> Option<u32> {
    let parts: Vec<&str> = ip.split('.').collect();
    if parts.len() != 4 {
        return None;
    }
    let mut val: u32 = 0;
    for p in parts {
        if p.is_empty() || p.len() > 3 || !p.bytes().all(|b| b.is_ascii_digit()) {
            return None;
        }
        let n: u32 = p.parse().ok()?;
        if n > 255 {
            return None;
        }
        val = (val << 8) | n;
    }
    Some(val)
}

fn in_cidr4(ip_int: u32, base_ip: &str, mask_bits: u32) -> bool {
    let base = match ipv4_to_int(base_ip) {
        Some(b) => b,
        None => return false,
    };
    let mask: u32 = if mask_bits == 0 {
        0
    } else {
        0xffff_ffffu32 << (32 - mask_bits)
    };
    (ip_int & mask) == (base & mask)
}

fn is_blocked_ipv4_str(ip: &str) -> bool {
    let n = match ipv4_to_int(ip) {
        Some(n) => n,
        None => return true,
    };
    in_cidr4(n, "0.0.0.0", 8)        // "this" network / unspecified
        || in_cidr4(n, "10.0.0.0", 8)        // private
        || in_cidr4(n, "100.64.0.0", 10)     // CGN
        || in_cidr4(n, "127.0.0.0", 8)       // loopback
        || in_cidr4(n, "169.254.0.0", 16)    // link-local (incl. metadata)
        || in_cidr4(n, "172.16.0.0", 12)     // private
        || in_cidr4(n, "192.0.0.0", 24)      // IETF protocol assignments
        || in_cidr4(n, "192.168.0.0", 16)    // private
        || in_cidr4(n, "198.18.0.0", 15)     // benchmarking
        || in_cidr4(n, "224.0.0.0", 4)       // multicast
        || in_cidr4(n, "240.0.0.0", 4) // reserved
}

/// True if an IP literal is private/loopback/link-local/ULA/cloud-metadata and
/// must NOT be fetched. Handles IPv4, IPv6, and IPv4-mapped IPv6.
/// Returns true for non-IP input (refuse).
pub fn is_blocked_ip(ip: &str) -> bool {
    // Determine family. node's isIP returns 0 for non-IPs.
    if ip.parse::<Ipv4Addr>().is_ok() {
        return is_blocked_ipv4_str(ip);
    }
    if ip.parse::<Ipv6Addr>().is_ok() {
        let lower = ip.to_lowercase();
        // IPv4-mapped (::ffff:a.b.c.d) — classify the v4 part.
        if let Some(rest) = lower.strip_prefix("::ffff:") {
            if rest.contains('.') {
                return is_blocked_ipv4_str(rest);
            }
        }
        if lower == "::1" || lower == "::" {
            return true; // loopback / unspecified
        }
        if lower.starts_with("fe80") {
            return true; // link-local
        }
        if lower.starts_with("fc") || lower.starts_with("fd") {
            return true; // unique local (ULA)
        }
        if lower.starts_with("ff") {
            return true; // multicast
        }
        return false;
    }
    // Not a valid IP => refuse.
    true
}

/// Outcome of `validate_outbound_url`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum OutboundCheck {
    Ok,
    BadScheme,
    BlockedIp,
    LocalHost,
}

/// Validate a URL's scheme/host shape before any network use. Pure.
/// Rejects non-http(s) schemes, blocked IP literals, and obvious local names.
pub fn validate_outbound_url(url: &url::Url) -> OutboundCheck {
    if url.scheme() != "http" && url.scheme() != "https" {
        return OutboundCheck::BadScheme;
    }
    let host = url.host_str().unwrap_or("");
    // If the host is an IP literal, classify immediately.
    let is_ip = host.parse::<Ipv4Addr>().is_ok() || host.parse::<Ipv6Addr>().is_ok();
    if is_ip && is_blocked_ip(host) {
        return OutboundCheck::BlockedIp;
    }
    let lower_host = host.to_lowercase();
    if lower_host == "localhost"
        || lower_host.ends_with(".localhost")
        || lower_host.ends_with(".local")
    {
        return OutboundCheck::LocalHost;
    }
    OutboundCheck::Ok
}

/// True if a destination is safe to fetch as far as the pure classifiers can
/// tell (scheme ok, not a local name, not a blocked IP literal). Live DNS
/// resolution is the worker's responsibility.
pub fn is_safe_destination(url: &url::Url) -> bool {
    validate_outbound_url(url) == OutboundCheck::Ok
}

#[cfg(test)]
mod tests {
    use super::*;
    use url::Url;

    #[test]
    fn blocks_private_and_loopback_ipv4() {
        assert!(is_blocked_ip("127.0.0.1"));
        assert!(is_blocked_ip("10.1.2.3"));
        assert!(is_blocked_ip("172.16.0.1"));
        assert!(is_blocked_ip("192.168.1.1"));
        assert!(is_blocked_ip("169.254.169.254")); // cloud metadata
        assert!(is_blocked_ip("100.64.0.1")); // CGN
        assert!(is_blocked_ip("0.0.0.0"));
    }

    #[test]
    fn allows_public_ipv4() {
        assert!(!is_blocked_ip("8.8.8.8"));
        assert!(!is_blocked_ip("1.1.1.1"));
        assert!(!is_blocked_ip("93.184.216.34"));
    }

    #[test]
    fn handles_ipv6() {
        assert!(is_blocked_ip("::1"));
        assert!(is_blocked_ip("::"));
        assert!(is_blocked_ip("fe80::1"));
        assert!(is_blocked_ip("fc00::1"));
        assert!(is_blocked_ip("fd12::1"));
        assert!(is_blocked_ip("ff02::1"));
        assert!(is_blocked_ip("::ffff:127.0.0.1")); // ipv4-mapped loopback
        assert!(!is_blocked_ip("2606:4700:4700::1111")); // public
    }

    #[test]
    fn refuses_non_ip_input() {
        assert!(is_blocked_ip("not-an-ip"));
        assert!(is_blocked_ip(""));
    }

    #[test]
    fn validate_outbound_branches() {
        assert_eq!(
            validate_outbound_url(&Url::parse("https://example.com/").unwrap()),
            OutboundCheck::Ok
        );
        assert_eq!(
            validate_outbound_url(&Url::parse("ftp://example.com/").unwrap()),
            OutboundCheck::BadScheme
        );
        assert_eq!(
            validate_outbound_url(&Url::parse("http://127.0.0.1/").unwrap()),
            OutboundCheck::BlockedIp
        );
        assert_eq!(
            validate_outbound_url(&Url::parse("http://localhost/").unwrap()),
            OutboundCheck::LocalHost
        );
        assert_eq!(
            validate_outbound_url(&Url::parse("http://foo.local/").unwrap()),
            OutboundCheck::LocalHost
        );
    }

    #[test]
    fn is_safe_destination_matches() {
        assert!(is_safe_destination(&Url::parse("https://example.com").unwrap()));
        assert!(!is_safe_destination(&Url::parse("http://10.0.0.1").unwrap()));
    }
}
